code/game/mecha/equipment/tools/janitor_tools.dm

```diff
@@ -0,0 +1,377 @@
+// Mecha mop, light replacer, mecha spray, garbage bag
+
+/obj/item/mecha_parts/mecha_equipment/janitor
+
+/obj/item/mecha_parts/mecha_equipment/janitor/can_attach(obj/mecha/janitor/M)
+	if(..() && istype(M))
+		return TRUE
+
+// Mop
+
+#define MOP_SOUND_CD 2 SECONDS // How many seconds before the mopping sound triggers again
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_mop
+	name = "WLLY mega mop"
+	desc = "An upsized advanced mop, designed for use in exosuits."
+	icon_state = "mecha_mop"
+	equip_cooldown = 15
+	energy_drain = 1
+	range = MECHA_MELEE | MECHA_RANGED
+	/// When the mopping sound was last played.
+	var/mop_sound_cooldown
+	/// How fast does this mop?
+	var/mop_speed = 20
+	/// Toggle for refilling itself
+	var/refill_enabled = TRUE
+	/// Rate per process() tick mop refills itself
+	var/refill_rate = 5
+	/// Power use per process to refill reagents
+	var/refill_cost = 10
+	/// What reagent to refill with
+	var/refill_reagent = "water"
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_mop/New()
+	create_reagents(1000)
+	reagents.add_reagent("water", 1000)
+	START_PROCESSING(SSobj, src)
+	..()
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_mop/emag_act(mob/user)
+	. = ..()
+	emagged = TRUE
+	to_chat(user, "<span class='notice'>You short out the automatic watering system on [src].</span>")
+	reagents.clear_reagents()
+	refill_reagent = "lube"
+	refill_cost = 50
+
+/obj/item/mecha_parts/mecha_equipment/janitor/mega_mop/action(atom/target)
+	if(istype(target, /obj/structure/reagent_dispensers/watertank) && get_dist(chassis,target) <= 1)
+		var/obj/structure/reagent_dispensers/watertank/WT = target
+		WT.reagents.trans_to(src, 1000)
+		occupant_message("<span class='notice'>Mop refilled.</span>")
+		playsound(chassis, 'sound/effects/refill.ogg', 50, TRUE, -6)
+	else
+		if(get_dist(chassis, target) > 2)
+			return
+		if(reagents.total_volume > 0)
+			if(world.time > mop_sound_cooldown)
+				playsound(loc, pick('sound/weapons/mopping1.ogg', 'sound/weapons/mopping2.ogg'), 30, TRUE, -1)
+				mop_sound_cooldown = world.time + MOP_SOUND_CD
+			// 3x3 mopping area
+			var/turf/target_turf = get_turf(target)
+			if(!istype(target_turf) || iswallturf(target_turf))
+				return
+			for(var/turf/current_target_turf in view(1, target))
+				INVOKE_ASYNC(current_target_turf, TYPE_PROC_REF(/atom, cleaning_act), chassis.occupant, src, mop_speed, "mop", ".")
```

Comment:
The issue with using a do_after in this case is that it would then clean each tile sequentially instead of simultaneously, which circumvents the entire use of the module's ability to clean a 3x3 area.

Reply:
I reworked the logic to have a do_after before the for loop, and added a parameter to the cleaning_act proc to prevent it from flooding the chat with "you mopped the floor" and allow it to do all nine tiles simultaneously by skipping the built-in do_after.